src/inspector_socket.cc

```diff
@@ -0,0 +1,762 @@
+                                 int status, const uv_buf_t* buf) {
+  if (inspector->ws_state->read_cb) {
+    inspector->ws_state->read_cb(
+        reinterpret_cast<uv_stream_t*>(&inspector->client), status, buf);
+  }
+}
+
+static void shutdown_complete(inspector_socket_t* inspector) {
+  if (inspector->ws_state->close_cb) {
+    inspector->ws_state->close_cb(inspector, 0);
+  }
+  close_connection(inspector);
+}
+
+static void on_close_frame_written(uv_write_t* write, int status) {
+  inspector_socket_t* inspector =
+      reinterpret_cast<inspector_socket_t*>(write->handle->data);
+  write_request_cleanup(write, status);
+  inspector->ws_state->close_sent = true;
+  if (inspector->ws_state->received_close) {
+    shutdown_complete(inspector);
+  }
+}
+
+static void close_frame_received(inspector_socket_t* inspector) {
+  inspector->ws_state->received_close = true;
+  if (!inspector->ws_state->close_sent) {
+    invoke_read_callback(inspector, 0, 0);
+    write_to_client(inspector, CLOSE_FRAME, sizeof(CLOSE_FRAME),
+                    on_close_frame_written);
+  } else {
+    shutdown_complete(inspector);
+  }
+}
+
+static int parse_ws_frames(inspector_socket_t* inspector, size_t len) {
+  int bytes_consumed = 0;
+  char* output = nullptr;
+  bool compressed = false;
+
+  ws_decode_result r =  decode_frame_hybi17(inspector->buffer,
+                                            len, true /* client_frame */,
+                                            &bytes_consumed, &output,
+                                            &compressed);
+  // Compressed frame means client is ignoring the headers and misbehaves
+  if (compressed || r == FRAME_ERROR) {
+    invoke_read_callback(inspector, UV_EPROTO, nullptr);
+    close_connection(inspector);
+    bytes_consumed = 0;
+  } else if (r == FRAME_CLOSE) {
+    close_frame_received(inspector);
+    bytes_consumed = 0;
+  } else if (r == FRAME_OK && inspector->ws_state->alloc_cb
+             && inspector->ws_state->read_cb) {
+    uv_buf_t buffer;
+    size_t len = strlen(output) + 1;
+    inspector->ws_state->alloc_cb(
+        reinterpret_cast<uv_handle_t*>(&inspector->client),
+        len, &buffer);
+    CHECK_GE(len, buffer.len);
+    strncpy(buffer.base, output, len);
+    invoke_read_callback(inspector, len, &buffer);
+  }
+  free(output);
+  return bytes_consumed;
+}
+
+static void prepare_buffer(uv_handle_t* stream, size_t len, uv_buf_t* buf) {
+  inspector_socket_t* inspector =
+      reinterpret_cast<inspector_socket_t*>(stream->data);
+
+  if (len > (inspector->buffer_size - inspector->data_len)) {
+    int new_size = (inspector->data_len + len + BUFFER_GROWTH_CHUNK_SIZE - 1) /
+                   BUFFER_GROWTH_CHUNK_SIZE *
+                   BUFFER_GROWTH_CHUNK_SIZE;
+    inspector->buffer_size = new_size;
+    inspector->buffer = reinterpret_cast<char*>(realloc(inspector->buffer,
+                                        inspector->buffer_size));
```

Comment:
Ditto, more or less.  Is there a reason you can't turn `inspector->buffer` into a `std::vector<char>`?